crates/cargo-bsp/src/project_model/sources.rs

```diff
@@ -0,0 +1,239 @@
+            let temp_file = File::create(temp_file_path.clone()).unwrap();
+
+            let cases = vec![
+                TestCase {
+                    path: temp_dir_path.clone(),
+                    expected: SourceItem {
+                        uri: file_uri(temp_dir_path.to_str().unwrap()),
+                        kind: SourceItemKind::Directory,
+                        generated: false,
+                    },
+                },
+                TestCase {
+                    path: temp_file_path.clone(),
+                    expected: SourceItem {
+                        uri: file_uri(temp_file_path.to_str().unwrap()),
+                        kind: SourceItemKind::File,
+                        generated: false,
+                    },
+                },
+            ];
+
+            for case in cases {
+                let source_item = create_source_item(case.path);
+                assert_eq!(source_item, case.expected);
+                assert!(source_item.uri.starts_with("file://"));
+            }
+
+            drop(temp_file);
+            temp_dir.close().unwrap();
+        }
+    }
+
+    mod get_all_rs_files_in_dir_test {
+        use std::collections::HashSet;
+        use std::path::Path;
+
+        use crate::project_model::sources::get_all_rs_files_in_dir;
+
+        use super::*;
+
+        struct TestCase {
+            dir_path: String,
+            expected: HashSet<PathBuf>,
+        }
+
+        const RUST_FILE_NAMES: [&str; 3] = ["test1.rs", "test2.rs", "test3.rs"];
+        const NOT_RUST_FILE_NAMES: [&str; 3] = ["test1.txt", "test4", "test5.rs.java"];
+
+        fn create_files(files_names: &[&str], dir: &Path) -> (HashSet<PathBuf>, Vec<File>) {
+            let files_paths = files_names
+                .iter()
+                .map(|name| dir.join(name))
+                .collect::<Vec<PathBuf>>();
+            let files = files_paths
+                .iter()
+                .map(|path| File::create(path.clone()).unwrap())
+                .collect::<Vec<File>>();
+
+            (files_paths.into_iter().collect(), files)
+        }
+
+        fn create_files_in_dirs(
+            files_names: &[&str],
+            dirs: &[TempDir],
+        ) -> (Vec<HashSet<PathBuf>>, Vec<Vec<File>>) {
+            dirs.iter()
+                .map(|dir| create_files(files_names, dir.path()))
+                .unzip()
+        }
+
+        fn create_dirs() -> Vec<TempDir> {
+            let dir1 = tempdir().unwrap();
+            let dir2 = tempdir_in(dir1.path()).unwrap();
+            let dir3 = tempdir_in(dir2.path()).unwrap();
+            let dir4 = tempdir_in(dir1.path()).unwrap();
+            let dir5 = tempdir_in(dir4.path()).unwrap();
+
+            vec![dir1, dir2, dir3, dir4, dir5]
+        }
+
+        #[test]
+        fn test() {
+            let dirs = create_dirs();
+
+            let (rust_files_paths, rust_files) =
+                create_files_in_dirs(&RUST_FILE_NAMES, &dirs.as_slice()[..=3]);
```

Comment:
to slicowanie się wydaje bardzo niewygodne i nieczytelne, to lepiej już
```suggestion
                create_files_in_dirs(&RUST_FILE_NAMES, &[dir_root, dir_root_a, dir_root_a_b, dir_root_b]);
```